Source/WebCore/dom/Element.cpp

```diff
@@ -5421,6 +5422,41 @@ void Element::setContentRelevancy(OptionSet<ContentRelevancy> contentRelevancy)
     ensureElementRareData().setContentRelevancy(contentRelevancy);
 }
 
+bool Element::checkVisibility(const CheckVisibilityOptions& options)
+{
+    document().updateStyleIfNeeded();
+
+    auto* style = computedStyle();
+
+    // Disconnected node, not rendered.
+    if (!style)
+        return false;
+
+    // See https://github.com/w3c/csswg-drafts/issues/9478.
+    if (style->display() == DisplayType::Contents)
+        return false;
+
+    if (options.checkVisibilityCSS && style->visibility() != Visibility::Visible)
+        return false;
+
+    for (RefPtr ancestor = this; ancestor; ancestor = ancestor->parentElementInComposedTree()) {
```

Comment:
The spec says to check the shadow-including ancestors of "this", not "in flat tree". They're different. In particular, a node assigned to a slot will have its parent node as the parent in shadow-including ancestors but will have its assigned slot as the part in flat tree.

Reply:
I think the flat tree is correct here, since we want to match CSS rendering. This is in fact what Chrome implements. I'll file a CSSWG issue to get this corrected.

Reply:
We should definitely add a test for this.

Reply:
I filed https://github.com/w3c/csswg-drafts/issues/9486 , I'll add a test tomorrow